### PR TITLE
CMakeLists.txt fixes

### PR DESCRIPTION
**Pull request by sidhax**
This fixes CMakeLists.txt and remove unwanted dependencies for IAITO.

### CODE REVIEW

**Comment by thestr4ng3r** [comment]
Sorry, I didn't notice you made this pr here. Here's my code review: https://github.com/thestr4ng3r/iaito/pull/1

File: src/CMakeLists.txt
Comment (thestr4ng3r) [inline src/CMakeLists.txt:11]: Please tell me if I'm wrong, but I think CheckCXXCompilerFlag and GenerateExportHeader are unused.

Also, why would you use set instead of list(APPEND ...) for CMAKE_MODULE_PATH? And please put  ${CMAKE_CURRENT_SOURCE_DIR}/cmake in quotes.

File: src/CMakeLists.txt
Comment (thestr4ng3r) [inline src/CMakeLists.txt:18]: FindRadare2.cmake (written by me) does not check the version, so this option will be ignored.

It might be helpful to add that functionality to the find module though, so people can't compile against any old versions.

File: src/CMakeLists.txt
Comment (thestr4ng3r) [inline src/CMakeLists.txt:24]: Please do not set CMAKE_INSTALL_PREFIX. This should be done by the user when calling cmake.

File: src/CMakeLists.txt
Comment (thestr4ng3r) [inline src/CMakeLists.txt:25]: No need to explicitly specify `${CMAKE_INSTALL_PREFIX}` here. if the path for `install(...)` is relative, it is automatically treated as relative to the install prefix.

File: src/CMakeLists.txt
Comment (thestr4ng3r) [inline src/CMakeLists.txt:28]: I'm not an expert on the whole moc thing, but why exactly would you like to remove AUTOMOC and AUTORCC? Wouldn't that just require qt5_wrap_* to be called manually?

File: src/CMakeLists.txt
Comment (thestr4ng3r) [inline src/CMakeLists.txt:34]: `link_directories(${RADARE2_LIBRARY_DIRS})` is required in some cases.

File: src/CMakeLists.txt
Comment (thestr4ng3r) [inline src/CMakeLists.txt:65]: Please never set compiler flags directly if not really necessary. For C++11, only `set(CMAKE_CXX_STANDARD 11)` should be used. Or do you have a specific reason to not use it?

And if you really want to set CMAKE_CXX_FLAGS manually, only append it and do not overwrite it completely unless you explicitly want to do that.

Also, `CMAKE_CXX_COMPILER_IS_GNUCXX` and `CMAKE_CXX_COMPILER_IS_CLANGXX` do not exist. I think it's best to check `CMAKE_CXX_COMPILER_ID` as it was before.

File: src/CMakeLists.txt
Comment (thestr4ng3r) [inline src/CMakeLists.txt:142]: Please leave the whole stuff from `parse_qmake_pro` in. I know, what QMakeProParse.cmake does is some black magic (I have written that script myself), but the reason I made it is that you don't have to explicitly specify the source files in CMakeLists.txt. Because most people use qmake, they will only update iaito.pro if a filename changes and thus break CMake.

**Review by thestr4ng3r — CHANGES_REQUESTED** [review]

File: src/CMakeLists.txt
Comment (sidhax) [reply src/CMakeLists.txt:11]: 1. CheckCXXCompilerFlag, I had something in mind to check for compiler flags but this is in wip.
2. GenerateExportHeader: will remove this yes you are right this is not used

File: src/CMakeLists.txt
Comment (sidhax) [reply src/CMakeLists.txt:18]: yes also code in main file checks for version of radare2 which is overhead.
1. Should not compile at first if it doesnt have radare2 version as required.
2. For distribution packaging radare2 version to checked as iaito's runtime dependecy instead of complaining after install.

* This may be different for users who are never going to use distribution packaging and just use it from git. Decision entirely depends on project maintainers.

File: src/CMakeLists.txt
Comment (sidhax) [reply src/CMakeLists.txt:24]: Yes this a dirty hack I had to set it because install targets wont work without it.

File: src/CMakeLists.txt
Comment (sidhax) [reply src/CMakeLists.txt:25]: Yes this is a dirty hack. I had to set it because install targets wont work without it. Also I mentioned clearly in PR this is in WIP

File: src/CMakeLists.txt
Comment (sidhax) [reply src/CMakeLists.txt:34]: find_package(Radare2 1.4.0 REQUIRED)
include_directories(${RADARE2_INCLUDE_DIRS})  still are part of file please recheck. Only link_directories(${RADARE2_LIBRARY_DIRS}) was removed because it is anyways linked afterwards.

File: src/CMakeLists.txt
Comment (sidhax) [reply src/CMakeLists.txt:65]: set(CMAKE_CXX_STANDARD 11) and -std=c++11 are same afaik.
well CMAKE_CXX_COMPILER_IS_GNUCXX do exist https://cmake.org/cmake/help/latest/variable/CMAKE_COMPILER_IS_GNUCXX.html.

CMAKE_COMPILER_IS_CLANGCXX do seem to exist but I also see typo in my PR. But I am OK if it isn't required

File: src/CMakeLists.txt
Comment (sidhax) [reply src/CMakeLists.txt:142]: hm yes if project maintainers want to keep .pro (project) file as main source to be updated.

File: src/CMakeLists.txt
Comment (thestr4ng3r) [reply src/CMakeLists.txt:18]: I see what you mean and I agree that it makes sense to check for the version when building.

But the problem is `find_package(Radare2 ...)` runs the FindRadare2.cmake script inside this repo under src/cmake and this script simply does not check the version at the current state.
So as long as this script is not adjusted to do so, the version should not be specified here, otherwise it makes people assume the version is checked which is not true.

File: src/CMakeLists.txt
Comment (thestr4ng3r) [reply src/CMakeLists.txt:24]: Can you share more details about what exactly would not work?

File: src/CMakeLists.txt
Comment (thestr4ng3r) [reply src/CMakeLists.txt:34]: `link_directories(${RADARE2_LIBRARY_DIRS})` is still needed, because `RADARE2_LIBRARIES` not always contains the full absolute paths for the libraries. On anything except Windows, FindRadare2.cmake use pkg-config to find radare, which sets `RADARE2_LIBRARIES` to something like `r_core r_bin ...` and `RADARE2_LIBRARY_DIRS` to the library dir.

File: src/CMakeLists.txt
Comment (sidhax) [reply src/CMakeLists.txt:18]: Can try to fix it.

File: src/CMakeLists.txt
Comment (sidhax) [reply src/CMakeLists.txt:24]: This will fail if CMAKE_INSTALL_PREFIX is not set , install(TARGETS iaito RUNTIME DESTINATION ${BIN_INSTALL_DIR})

File: src/CMakeLists.txt
Comment (thestr4ng3r) [reply src/CMakeLists.txt:65]: `set(CMAKE_CXX_STANDARD 11)` and `-std=c++11` are only the same if the compiler supports `-std=c++11`. CMake already has the functionality to set the C++ standard for all supported compilers automatically using `CMAKE_CXX_STANDARD`, so one should also take advantage of that instead of manually specifying flags.

What I meant by `CMAKE_CXX_COMPILER_IS_GNUCXX` does not exist is that it's called `CMAKE_COMPILER_IS_GNUCXX`. Clang should also definitely be supported. But even in the documentation in the link you sent, they suggest to use `CMAKE_CXX_COMPILER_ID` instead of `CMAKE_COMPILER_IS_GNUCXX`.

File: src/CMakeLists.txt
Comment (thestr4ng3r) [reply src/CMakeLists.txt:142]: If you ask me, we could throw out all qmake stuff completely from the project, but most people here prefer it, so I think this is the best compromise.

File: src/CMakeLists.txt
Comment (thestr4ng3r) [reply src/CMakeLists.txt:11]: I see. But please remove the includes for the pr if they are not used.

File: src/CMakeLists.txt
Comment (sidhax) [reply src/CMakeLists.txt:142]: I haven't checked completely but to generate project file 'qmake-qt5 -project' can be run in root of the project folder. Well if project is big it does take some time to parse and go through all files.

File: src/CMakeLists.txt
Comment (sidhax) [reply src/CMakeLists.txt:11]: will do

File: src/widgets/functionswidget.h
Comment (ballessay) [inline src/widgets/functionswidget.h]: Could be further reduced to:
```
#include "dockwidget.h"
#include "qrcore.h"

#include <QSortFilterProxyModel>
#include <QTreeView>
```

**Review by ballessay — COMMENTED** [review]
Would have been nice if you split up the PR, one for code and one for cmake changes. 

Code changes look ok to me.

File: src/widgets/functionswidget.h
Comment (sidhax) [reply src/widgets/functionswidget.h]: Will split up PR. Thank you.

File: src/CMakeLists.txt
Comment (sidhax) [reply src/CMakeLists.txt:142]: I agree with you about leaving the part which parses Qt project file for Iaito; considering this is in development and likely to have lot of changes, I will flip it back in cmake file.

Comment (ballessay) [reply src/CMakeLists.txt:142]: Well `qmake -project` will just generate a minimal .pro (TEMPLATE=app, TARGET=$PWD..) with the sources, headers, resources... . Dependencies to other libraries, INSTALL-targets and so on have to be written/integrated manually.

File: src/CMakeLists.txt
Comment (thestr4ng3r) [reply src/CMakeLists.txt:24]: Ah, I see! Then I guess that's because `${CMAKE_INSTALL_PREFIX}/bin` will become `/bin`.
Have you tried the following?
```
set(BIN_INSTALL_DIR bin)
install(TARGETS iaito RUNTIME DESTINATION ${BIN_INSTALL_DIR})
```
So install will only receive "bin" without the slash in front of it and use `CMAKE_INSTALL_PREFIX` automatically.

But wouldn't `CMAKE_INSTALL_PREFIX` be set to /usr/local by default?

File: src/CMakeLists.txt
Comment (sidhax) [reply src/CMakeLists.txt:65]: Will flip it back

File: src/CMakeLists.txt
Comment (sidhax) [reply src/CMakeLists.txt:24]: It should be

iaito -> /usr/bin/iaito
{libs} x86_64 -> /usr/lib64
{libs} x86 -> /usr/lib
{other miscellaneous data files} -> /usr/share/iaito
{licence} -> /usr/share/licenses/iaito/
{themes} -> /usr/share/iaito/themes
{icons} -> /usr/share/icons/  or /usr/share/icons/iaito
logo -> /usr/share/iaito/iaito.png // this needs fix as commented in qmake project file
resource qrc files -> /usr/share/iaito/images

Comment (sidhax) [reply src/CMakeLists.txt:24]: afaik default prefix path is /usr , but this might be different on different distributions. This can be overridden by using -DCMAKE_INSTALL_PREFIX=/usr/local or any other place when running cmake.

File: src/CMakeLists.txt
Comment (sidhax) [reply src/CMakeLists.txt:142]: Yes, you are correct about this. I can actually write script to expand or transform into CMakeLists.txt.

File: src/widgets/functionswidget.h
Comment (sidhax) [reply src/widgets/functionswidget.h]: moved to https://github.com/hteso/iaito/pull/171

**Comment by radare** [comment]
ping.

**Comment by radare** [comment]
right now the cmake build is not working for me, so i think that merging this will at least improve the situation

**Comment by thestr4ng3r** [comment]
Please don't merge it yet (see my review).

What exactly is not working for you? I use cmake for Iaito all the time and it works fine for me.

**Comment by trufae** [comment]
It was working a month ago when i tested it. I tried yesterday and it was just failing. Cant remember where now

**Comment by thestr4ng3r** [comment]
It might be required to rerun cmake to update with changes in iaito.pro. I think I will add a small fix to trigger that automatically.

**Comment by mrexodia** [comment]
What's going on with this pull request? Is it ready to be merged?

**Comment by thestr4ng3r** [comment]
No, please take a look at my review.

**Comment by mrexodia** [comment]
@sidhax what is this supposed to fix exactly?

File: src/CMakeLists.txt
Comment (radare) [reply src/CMakeLists.txt:11]: Eta wen

Comment (radare) [reply src/CMakeLists.txt:11]: ping

File: src/CMakeLists.txt
Comment (radare) [reply src/CMakeLists.txt:18]: wen

**Comment by radare** [comment]
pls fix conflicts

**Comment by xarkes** [comment]
ping ?

**Comment by radare** [comment]
PING

**Comment by xarkes** [comment]
Please resolve conflicts or close PR

**Comment by sidhax** [comment]
closing pull request

**Pull request closed without merging**
